internal/component/database_observability/mysql/collector/query_sample.go

```diff
@@ -0,0 +1,290 @@
+       statements.MAX_TOTAL_MEMORY
+FROM performance_schema.events_statements_history AS statements
+JOIN performance_schema.global_status
+WHERE statements.sql_text IS NOT NULL
+  AND global_status.variable_name = 'UPTIME'
+  AND statements.CURRENT_SCHEMA NOT IN ('mysql', 'performance_schema', 'sys', 'information_schema')
+  AND statements.TIMER_END >= ?;`
+
+type QuerySampleArguments struct {
+	DB              *sql.DB
+	InstanceKey     string
+	CollectInterval time.Duration
+	EntryHandler    loki.EntryHandler
+	UseTiDBParser   bool
+
+	Logger log.Logger
+}
+
+type QuerySample struct {
+	dbConnection    *sql.DB
+	instanceKey     string
+	collectInterval time.Duration
+	entryHandler    loki.EntryHandler
+	sqlParser       parser.Parser
+
+	logger  log.Logger
+	running *atomic.Bool
+	ctx     context.Context
+	cancel  context.CancelFunc
+
+	lastSampleSeenTimestamp float64
+}
+
+func NewQuerySample(args QuerySampleArguments) (*QuerySample, error) {
+	c := &QuerySample{
+		dbConnection:    args.DB,
+		instanceKey:     args.InstanceKey,
+		collectInterval: args.CollectInterval,
+		entryHandler:    args.EntryHandler,
+		logger:          log.With(args.Logger, "collector", QuerySampleName),
+		running:         &atomic.Bool{},
+	}
+
+	if args.UseTiDBParser {
+		c.sqlParser = parser.NewTiDBSqlParser()
+	} else {
+		c.sqlParser = parser.NewXwbSqlParser()
+	}
+
+	return c, nil
+}
+
+func (c *QuerySample) Name() string {
+	return QuerySampleName
+}
+
+func (c *QuerySample) Start(ctx context.Context) error {
+	level.Debug(c.logger).Log("msg", QuerySampleName+" collector started")
+
+	c.running.Store(true)
+	ctx, cancel := context.WithCancel(ctx)
+	c.ctx = ctx
+	c.cancel = cancel
+
+	if err := c.setLastSampleSeenTimestamp(c.ctx); err != nil {
+		level.Error(c.logger).Log("msg", "failed to set last sample seen timestamp", "err", err)
+		return err
+	}
+
+	go func() {
+		defer func() {
+			c.Stop()
+			c.running.Store(false)
+		}()
+
+		ticker := time.NewTicker(c.collectInterval)
+
+		for {
+			if err := c.fetchQuerySamples(c.ctx); err != nil {
+				level.Error(c.logger).Log("msg", "collector error", "err", err)
+			}
+
+			select {
+			case <-c.ctx.Done():
+				return
+			case <-ticker.C:
+				// continue loop
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (c *QuerySample) Stopped() bool {
+	return !c.running.Load()
+}
+
+// Stop should be kept idempotent
+func (c *QuerySample) Stop() {
+	c.cancel()
+}
+
+// setLastSampleSeenTimestamp queries the database for the latest sample seen timestamp so that upon startup we don't collect
+// samples that may have been collected previously.
+func (c *QuerySample) setLastSampleSeenTimestamp(ctx context.Context) error {
+	rs, err := c.dbConnection.QueryContext(ctx, selectLatestTimerEnd)
+	if err != nil {
+		return fmt.Errorf("failed to fetch last sample seen timestamp: %w", err)
+	}
+	defer rs.Close()
+
+	for rs.Next() {
+		if err := rs.Err(); err != nil {
+			return fmt.Errorf("failed to iterate rows: %w", err)
+		}
```

Comment:
This check should be _after_ the loop (i.e. like on line 284).

Reply:
Addressed in 0d04fb23081c7112428d0035e7f738769e08bfdf